README
Bitcoin::RPC::Client

This module implements in PERL the methods that are currently part of the
Bitcoin Core RPC client calls (bitcoin-cli).The method names and parameters
are identical between the Bitcoin Core API and this module. This is done for
consistency so that a developer only has to reference one manual:
https://bitcoin.org/en/developer-reference#getinfo

The intention is to have a lightweight pure PERL implementation of the Bitcoin
Core API.

PREREQUISITES 
   Moo 
   JSON::RPC::Client

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Bitcoin::RPC::Client 

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Bitcoin-RPC-Client

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Bitcoin-RPC-Client

    CPAN Ratings
        http://cpanratings.perl.org/d/Bitcoin-RPC-Client

    Search CPAN
        http://search.cpan.org/dist/Bitcoin-RPC-Client/

AVAILABILITY

The latest branch is avaiable from Github:
   https://github.com/whinds84/Bitcoin-RPC-Client.git
   

LICENSE AND COPYRIGHT

Copyright (C) 2016 Wesley Hinds

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Bitcoin::RPC::Client 

0.05    2017-1-16/14
        - New Debug flag dumps HTTP data
        - Print to STDERR error strings same as bitcoin-cli
        - Support for boolean parameters 

0.04    2016-10-17/14
        Add warnings and strict.

0.03    2016-10-13/10
        Fixed broken test.

0.02    2016-10-12/15
        Added META files.

0.01    2016-10-12/10
        First version, released on an unsuspecting world.


lib/BTC.pm
package BTC;

use 5.008;
use Moo;
use JSON::RPC::Client;

use BTC::API;

our $VERSION  = '0.01';

has jsonrpc  => (is => "lazy", default => sub { "JSON::RPC::Client"->new });
has user     => (is => 'ro');
has password => (is => 'ro');
has host     => (is => 'ro');
has port     => (is => "lazy", default => 8332);
has timeout  => (is => "lazy", default => 20);

# SSL constructor options
has ssl      => (is => 'ro', default => 0);
has verify_hostname => (is => 'ro', default => 1);

sub AUTOLOAD {
   my $self   = shift;
   my $method = $BTC::AUTOLOAD;

   $method =~ s/.*:://;

   return if ($method eq 'DESTROY');
   # Thanks JSON::RPC::Client

   # Are we using SSL?
   my $uri = "http://";
   if ($self->ssl eq 1) {
      $uri = "https://";
   }
   my $url = $uri . $self->user . ":" . $self->password . "\@" . $self->host . ":" . $self->port;

   my $client = $self->jsonrpc;
   $client->ua->timeout($self->timeout); # Because bitcoind is slow
   # For self signed certs
   if ($self->verify_hostname eq 0) {
      $client->ua->ssl_opts( verify_hostname => 0 );
   }

   my @params = @_;
   my $obj = {
      method => $method,
      params => (ref $_[0] ? $_[0] : [@_]),
   };

   my $res = $client->call( $url, $obj );
   if($res) {
      if ($res->is_error) {
         print STDERR "error : ", $res->error_message->{message};
         return $res->error_message;
      }

      return $res->result;
   }

   # Else the service is down or incorrect
   print STDERR $client->status_line;

   # Check that method is even availabe in the API
   if (hasMethod($method)) { print "\n"; }
   # Print SYNTAX for API call
   if (failMethod($method)) { print "\n"; }

   return;
}

sub hasMethod {
   my ($method) = @_;

   # Check that method is even availabe in the API
   my $hasMethod = 0;
   foreach my $meth (@BTC::API::methods) {
      if (lc($meth) eq lc($method)) {
         $hasMethod = 1;
         last;
      }
   }

   if (!$hasMethod) {
      print STDERR "error : Method $method does not exist";
   }

   return 1;
}

sub failMethod {
   my ($method) = @_;

   foreach my $entry (@BTC::API::help) {
      if ($entry =~ /$method / || $entry =~ /$method$/) {
         print STDERR "error : usage: $entry";
         last;
      }
   }

   return 1;
}

1;

=pod

=head1 NAME

BTC - Bitcoin Core API RPCs

=head1 SYNOPSIS


   # Your bitcoind instances credentials
   $rpchost = "Your.IP.Address.Here";
   $rpcuser = "YourBitcoindRPCUserName";
   $rpcpassword = "YourBitcoindRPCPassword";
   # rpcport defaults to 8332

   # Create BTC object
   $btc = BTC->new(
      user     => $rpcuser,
      password => $rpcpassword,
      host     => $rpchost,
      ssl      => 1 # Optional
   );

   # Getting Data when a hash is returned
   $info    = $btc->getinfo;
   $balance = $info->{balance};
   print $balance;

  A person would need to know the JSON elements of
  the output. Ex.

   {
      "version" : 80100,
      "protocolversion" : 70001,
      "walletversion" : 60000,
      "balance" : 0.00720000,
      "blocks" : 253032,
      "connections" : 16,
      "proxy" : "",
      "difficulty" : 50810339.04827648,
      "testnet" : false,
      "keypoololdest" : 1365114158,
      "keypoolsize" : 101,
      "paytxfee" : 0.00500000,
      "errors" : ""
   }

   # Other functions that do not return a JSON object will have a scalar result
   $balance  = $btc->getbalance("yourAccountName");
   print $balance;

=head1 DESCRIPTION

This module implements in PERL the functions that are currently part of the
Bitcoin Core RPC client calls (bitcoin-cli).The function names and parameters
are identical between the Bitcoin Core API and this module. This is done for
consistency so that a developer only has to reference one manual:
https://bitcoin.org/en/developer-reference#getinfo

=head1 AUTHOR

Wesley Hinds wesley.hinds@swri.org

=head1 AVAILABILITY

The latest branch is avaiable from Github.

https://github.com/whinds84/BitcoinPM

=head1 LICENSE AND COPYRIGHT

Copyright 2016 Wesley Hinds.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut


lib/BitcoinPM.pm
#  BitcoinPM
#  Version 0.01 Copyright (c) 2015 Wesley Hinds <wesley.hinds@swri.org>
#  All rights reserved
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.
package BitcoinPM;

use 5.008;
use Moo;
use JSON;
use JSON::RPC::Client;

our $VERSION  = '0.01';

$ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} = 0;

has jsonrpc  => (is => "lazy", default => sub { "JSON::RPC::Client"->new });
has user     => (is => 'ro');
has password => (is => 'ro');
has host     => (is => 'ro');
has port     => (is => "lazy", default => 8332);

=cut
   client_operation. Aggregates the common RPC/JSON operations into
   one place.
=cut

sub client_operation{
   my $self = shift;
   my $url = "https://" . $self->user . ":" . $self->password . "\@" . $self->host . ":" . $self->port;

   my $obj = shift;

   my $client = $self->jsonrpc;

   my $res = $client->call( $url, $obj );

   if($res) {
      if ($res->is_error) {
         print STDERR "error : ", $res->error_message->{message};
         return $res->error_message;
      }
      return $res->result;
   }

   # Else the service is down or incorrect
   print STDERR $client->status_line;                                                                                                     
   return;
}

=cut
  getinfo
         Returns an object containing server information.

   {
      "version" : 80100,
      "protocolversion" : 70001,
      "walletversion" : 60000,
      "balance" : 0.00720000,
      "blocks" : 253032,                                                                                                                   
      "connections" : 16,                                                                                                                  
      "proxy" : "",                                                                                                                        
      "difficulty" : 50810339.04827648,                                                                                                    
      "testnet" : false,                                                                                                                   
      "keypoololdest" : 1365114158,                                                                                                        
      "keypoolsize" : 101,                                                                                                                 
      "paytxfee" : 0.00500000,                                                                                                             
      "errors" : ""                                                                                                                        
   } 
=cut

sub getinfo {
   my $self = shift;

   my $obj = {
      method  => 'getinfo',
   };

   return &client_operation($self,$obj);
}

=cut
   getaccount 'bitcoinaddress'
         Returns the account associated with the given address.

=cut

sub getaccount {

   my $self = shift;
   my $address = shift;

   my $obj = {
      method  => 'getaccount',
      params  => [ $address ]
   };

   return &client_operation($self,$obj);
}
=cut

   getbalance 'account'
         Returns the server's available balance, or the balance for 'account'.

=cut

sub getbalance {

   my $self = shift;
   my $account = shift;

   my $obj = {
      method  => 'getbalance',
      params  => [ $account ]
   };

   return &client_operation($self,$obj);
}

=cut

   getreceivedbyaccount 'account' ['minconf=1']
         Returns the total amount received by addresses associated with 'account' in transactions with at least ['minconf'] confirmations.

=cut

sub getreceivedbyaccount {

   my $self = shift;
   my $account = shift;
   my $minconf = shift;

   my $obj = {
      method  => 'getreceivedbyaccount',
      params  => [ $account ]
   };

   push $obj->{params},"minconf=$minconf" if ($minconf);

   return &client_operation($self,$obj);
} 

=cut                                                                                                                                      
                                                                                                                                          
   move <'fromaccount'> <'toaccount'> <'amount'> ['minconf=1'] ['comment']
              Moves funds between accounts.
                                                                                                                                          
=cut

sub move {
   my $self = shift;                                                                                                                       

   my $fromaccount = shift;
   my $toaccount = shift;
   my $amount = shift;
   my $minconf = shift;
   my $comment = shift;

   my $obj = {
      method  => 'move',
      params  => [ $fromaccount, $toaccount, $amount ]
   };
   # Add optional params
   push $obj->{params},"minconf=$minconf" if ($minconf);
   push $obj->{params}, $comment if ($comment);

   return &client_operation($self,$obj);
}

1;

__END__

=head1 NAME

BitcoinPM - access a local bitcoind with

=head1 SYNOPSIS

Creating the client

require("./include/BitcoinPM.inc");

# Create RPC connection object
$btcd = BitcoinPM->new(
    user     => 'rpcusername',
    password => 'rpcpassword',
    path     => '/path/to/bitcoind/binary',
  );

Getting Data

  $info    = $btcd->get_info;
  $balance = $info->{balance};

  A person would need to know the JSON elements of
  the output. Ex.

   {
      "version" : 80100,
      "protocolversion" : 70001,
      "walletversion" : 60000,
      "balance" : 0.00720000,
      "blocks" : 253032,
      "connections" : 16,
      "proxy" : "",
      "difficulty" : 50810339.04827648,
      "testnet" : false,
      "keypoololdest" : 1365114158,
      "keypoolsize" : 101,
      "paytxfee" : 0.00500000,
      "errors" : ""
   }

=head1 DESCRIPTION

BitcoinPM gives acces to an already running bitcoind (wallet) instance

Gives users and object oriented method of approaching the bitcoind in 
perl. Parses JSON output as well.

Old Way
sub get_account {
   my $self = shift;
   my $cmd  = $self->path . "-rpcssl -rpcconnect=" . $self->host . " -rpcuser=" . $self->user . " -rpcpassword=" . $self->password . " ";

   # Args
   my ($address) = @_; 

   # The bitcoin function
   $cmd    .= "getaccount '$address'";  

   my $result;
   $result = `$cmd 2>&1`; 

   # Do something if there's and error
   $result = &handle_error($result);

   return $result;
}

=cut


lib/Bitcoin/RPC/Client/API.pm
package Bitcoin::RPC::Client::API;

use strict;
use warnings;

our $VERSION  = '0.04';

our @methods = (
#==Blockchain}
qq{getbestblockhash},
qq{getblock},
qq{getblockchaininfo},
qq{getblockcount},
qq{getblockhash},
qq{getblockheader},
qq{getchaintips},
qq{getdifficulty},
qq{getmempoolinfo},
qq{getrawmempool},
qq{gettxout},
qq{gettxoutproof},
qq{gettxoutsetinfo},
qq{verifychain},
qq{verifytxoutproof},
#==Control}
qq{getinfo},
qq{help},
qq{stop},
#==Generating}
qq{generate},
qq{getgenerate},
qq{setgenerate},
#==Mining}
qq{getblocktemplate},
qq{getmininginfo},
qq{getnetworkhashps},
qq{prioritisetransaction},
qq{submitblock},
#==Network}
qq{addnode},
qq{clearbanned},
qq{disconnectnode},
qq{getaddednodeinfo},
qq{getconnectioncount},
qq{getnettotals},
qq{getnetworkinfo},
qq{getpeerinfo},
qq{listbanned},
qq{ping},
qq{setban},
#==Rawtransactions}
qq{createrawtransaction},
qq{decoderawtransaction},
qq{decodescript},
qq{fundrawtransaction},
qq{getrawtransaction},
qq{sendrawtransaction},
qq{signrawtransaction},
#==Util}
qq{createmultisig},
qq{estimatefee},
qq{estimatepriority},
qq{estimatesmartfee},
qq{estimatesmartpriority},
qq{validateaddress},
qq{verifymessage},
#==Wallet}
qq{abandontransaction},
qq{addmultisigaddress},
qq{backupwallet},
qq{dumpprivkey},
qq{dumpwallet},
qq{encryptwallet},
qq{getaccount},
qq{getaccountaddress},
qq{getaddressesbyaccount},
qq{getbalance},
qq{getnewaddress},
qq{getrawchangeaddress},
qq{getreceivedbyaccount},
qq{getreceivedbyaddress},
qq{gettransaction},
qq{getunconfirmedbalance},
qq{getwalletinfo},
qq{importaddress},
qq{importprivkey},
qq{importpubkey},
qq{importwallet},
qq{keypoolrefill},
qq{listaccounts},
qq{listaddressgroupings},
qq{listlockunspent},
qq{listreceivedbyaccount},
qq{listreceivedbyaddress},
qq{listsinceblock},
qq{listtransactions},
qq{listunspent},
qq{lockunspent},
qq{move},
qq{sendfrom},
qq{sendmany},
qq{sendtoaddress},
qq{setaccount},
qq{settxfee},
qq{signmessage},
);

our @help = (
#== Blockchain ==}
qq{getbestblockhash},
qq{getblock "hash" ( verbose )},
qq{getblockchaininfo},
qq{getblockcount},
qq{getblockhash index},
qq{getblockheader "hash" ( verbose )},
qq{getchaintips},
qq{getdifficulty},
qq{getmempoolinfo},
qq{getrawmempool ( verbose )},
qq{gettxout "txid" n ( includemempool )},
qq{gettxoutproof ["txid",...] ( blockhash )},
qq{gettxoutsetinfo},
qq{verifychain ( checklevel numblocks )},
qq{verifytxoutproof "proof"},
#== Control ==}
qq{getinfo},
qq{help ( "command" )},
qq{stop},
#== Generating ==}
qq{generate numblocks},
qq{getgenerate},
qq{setgenerate generate ( genproclimit )},
#== Mining ==}
qq{getblocktemplate ( "jsonrequestobject" )},
qq{getmininginfo},
qq{getnetworkhashps ( blocks height )},
qq{prioritisetransaction <txid> <priority delta> <fee delta>},
qq{submitblock "hexdata" ( "jsonparametersobject" )},
#== Network ==}
qq{addnode "node" "add|remove|onetry"},
qq{clearbanned},
qq{disconnectnode "node" },
qq{getaddednodeinfo dns ( "node" )},
qq{getconnectioncount},
qq{getnettotals},
qq{getnetworkinfo},
qq{getpeerinfo},
qq{listbanned},
qq{ping},
qq{setban "ip(/netmask)" "add|remove" (bantime) (absolute)},
#== Rawtransactions ==}
qq{createrawtransaction [{"txid":"id","vout":n},...] {"address":amount,"data":"hex",...} ( locktime )},
qq{decoderawtransaction "hexstring"},
qq{decodescript "hex"},
qq{fundrawtransaction "hexstring" includeWatching},
qq{getrawtransaction "txid" ( verbose )},
qq{sendrawtransaction "hexstring" ( allowhighfees )},
qq{signrawtransaction "hexstring" ( [{"txid":"id","vout":n,"scriptPubKey":"hex","redeemScript":"hex"},...] ["privatekey1",...] sighashtype )},
#== Util ==}
qq{createmultisig nrequired ["key",...]},
qq{estimatefee nblocks},
qq{estimatepriority nblocks},
qq{estimatesmartfee nblocks},
qq{estimatesmartpriority nblocks},
qq{validateaddress "bitcoinaddress"},
qq{verifymessage "bitcoinaddress" "signature" "message"},
#== Wallet ==}
qq{abandontransaction "txid"},
qq{addmultisigaddress nrequired ["key",...] ( "account" )},
qq{backupwallet "destination"},
qq{dumpprivkey "bitcoinaddress"},
qq{dumpwallet "filename"},
qq{encryptwallet "passphrase"},
qq{getaccount "bitcoinaddress"},
qq{getaccountaddress "account"},
qq{getaddressesbyaccount "account"},
qq{getbalance ( "account" minconf includeWatchonly )},
qq{getnewaddress ( "account" )},
qq{getrawchangeaddress},
qq{getreceivedbyaccount "account" ( minconf )},
qq{getreceivedbyaddress "bitcoinaddress" ( minconf )},
qq{gettransaction "txid" ( includeWatchonly )},
qq{getunconfirmedbalance},
qq{getwalletinfo},
qq{importaddress "address" ( "label" rescan p2sh )},
qq{importprivkey "bitcoinprivkey" ( "label" rescan )},
qq{importpubkey "pubkey" ( "label" rescan )},
qq{importwallet "filename"},
qq{keypoolrefill ( newsize )},
qq{listaccounts ( minconf includeWatchonly)},
qq{listaddressgroupings},
qq{listlockunspent},
qq{listreceivedbyaccount ( minconf includeempty includeWatchonly)},
qq{listreceivedbyaddress ( minconf includeempty includeWatchonly)},
qq{listsinceblock ( "blockhash" target-confirmations includeWatchonly)},
qq{listtransactions ( "account" count from includeWatchonly)},
qq{listunspent ( minconf maxconf  ["address",...] )},
qq{lockunspent unlock [{"txid":"txid","vout":n},...]},
qq{move "fromaccount" "toaccount" amount ( minconf "comment" )},
qq{sendfrom "fromaccount" "tobitcoinaddress" amount ( minconf "comment" "comment-to" )},
qq{sendmany "fromaccount" {"address":amount,...} ( minconf "comment" ["address",...] )},
qq{sendtoaddress "bitcoinaddress" amount ( "comment" "comment-to" subtractfeefromamount )},
qq{setaccount "bitcoinaddress" "account"},
qq{settxfee amount},
qq{signmessage "bitcoinaddress" "message"},
);

=pod

=head1 NAME

Bitcoin::RPC::Client::API - Bitcoin Core API RPCs

=head1 SYNOPSIS

   use Bitcoin::RPC::Client::API;

   # print all avaiable API calls
   foreach $method (@Bitcoin::RPC::Client::API::methods) {
      print $method;
   }

=head1 DESCRIPTION

A module that has lists of API commands and their syntax.
Generated from $ bitcoin-cli help
https://bitcoin.org/en/developer-reference

=head1 AUTHOR

Wesley Hinds wesley.hinds@gmail.com

=head1 AVAILABILITY

The latest branch is avaiable from Github.

https://github.com/whindsx/Bitcoin-RPC-Client.git

=head1 LICENSE AND COPYRIGHT

Copyright 2016 Wesley Hinds.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut


lib/Bitcoin/RPC/Client.pm
package Bitcoin::RPC::Client;

use 5.008;

use strict;
use warnings;

use Moo;
use JSON::RPC::Client;

use Bitcoin::RPC::Client::API;

our $VERSION  = '0.04';

has jsonrpc  => (is => "lazy", default => sub { "JSON::RPC::Client"->new });
has user     => (is => 'ro');
has password => (is => 'ro');
has host     => (is => 'ro');
has port     => (is => "lazy", default => 8332);
has timeout  => (is => "lazy", default => 20);
has syntax   => (is => "lazy", default => 0);
has debug    => (is => "lazy", default => 0);

# SSL constructor options
#  OpenSSL support has been removed from Bitcoin Core as of v0.12.0
#  but should work with older versions
has ssl      => (is => 'ro', default => 0);
has verify_hostname => (is => 'ro', default => 1);

sub AUTOLOAD {
   my $self   = shift;
   my $method = $Bitcoin::RPC::Client::AUTOLOAD;

   $method =~ s/.*:://;

   return if ($method eq 'DESTROY');
   # Thanks JSON::RPC::Client

   # Are we using SSL?
   my $uri = "http://";
   if ($self->ssl eq 1) {
      $uri = "https://";
   }
   my $url = $uri . $self->user . ":" . $self->password . "\@" . $self->host . ":" . $self->port;

   my $client = $self->jsonrpc;
   $client->ua->timeout($self->timeout); # Because bitcoind is slow

   # Turn on debugging for LWP::UserAgent
   if ($self->debug) {
      $client->ua->add_handler("request_send",  sub { shift->dump; return });
      $client->ua->add_handler("response_done", sub { shift->dump; return });
   } else {
      # We want to handle broken responses ourself
      $client->ua->add_handler("response_data", 
         sub { 
            my ($response, $ua, $h, $data) = @_;

            if ($response->is_error) {
               my $content = JSON->new->utf8->decode($data);

               print STDERR "error code: ";
               print STDERR $content->{error}->{code} . "\n";
               print STDERR "error message:\n";
               print STDERR $content->{error}->{message}; 
            }

            return;
         }
      );
   }

   # For self signed certs
   if ($self->verify_hostname eq 0) {
      $client->ua->ssl_opts( verify_hostname => 0 );
   }

   #my @params = @_;
   # ^ Should this be here?

   # Need to fix booleans.
   # JSON::true
   # But cant modify @_

   my $obj = {
      method => $method,
      params => (ref $_[0] ? $_[0] : [@_]),
   };

   my $res = $client->call( $url, $obj );
   if($res) {
      if ($res->is_error) {
         return $res->error_message;
      }

      return $res->result;
   }

   # Usage errors are on by default
   if ($self->syntax) {
      # Check that method is even availabe in the API
      hasMethod($method);

      # Print SYNTAX for API call
      failMethod($method);
   }

   return;
}

sub hasMethod {
   my ($method) = @_;

   # Check that method is even availabe in the API
   my $hasMethod = 0;
   foreach my $meth (@Bitcoin::RPC::Client::API::methods) {
      if (lc($meth) eq lc($method)) {
         $hasMethod = 1;
         last;
      }
   }

   if (!$hasMethod) {
      print STDERR "error : Method $method does not exist\n";
   }

   return 1;
}

sub failMethod {
   my ($method) = @_;

   foreach my $entry (@Bitcoin::RPC::Client::API::help) {
      if ($entry =~ /$method / || $entry =~ /$method$/) {
         print STDERR "error : usage: $entry\n";
         last;
      }
   }

   return 1;
}

1;

=pod

=head1 NAME

Bitcoin::RPC::Client - Bitcoin Core API RPCs

=head1 SYNOPSIS

   use Bitcoin::RPC::Client;

   # Create Bitcoin::RPC::Client object
   $btc = Bitcoin::RPC::Client->new(
      user     => "username",
      password => "p4ssword",
      host     => "127.0.0.1",
   );

   # Getting Data when a hash is returned
   $info    = $btc->getinfo;
   $balance = $info->{balance};
   print $balance;

   # A person would need to know the JSON elements of
   # the output. 
   # https://bitcoin.org/en/developer-reference#getinfo
   #
   # Ex.
   #{
   #   "version" : 80100,
   #   "protocolversion" : 70001,
   #   "walletversion" : 60000,
   #   "balance" : 0.00720000,
   #   "blocks" : 253032,
   #   "connections" : 16,
   #   "proxy" : "",
   #   "difficulty" : 50810339.04827648,
   #   "testnet" : false,
   #   "keypoololdest" : 1365114158,
   #   "keypoolsize" : 101,
   #   "paytxfee" : 0.00500000,
   #   "errors" : ""
   #}

   # Other functions that do not return a JSON object will have a scalar result
   $balance  = $btc->getbalance("yourAccountName");
   print $balance;

   # JSON::Boolean objects must be passed as boolean parameters
   $balance  = $btc->getbalance("yourAccountName", 1, JSON::true);
   print $balance;

=head1 DESCRIPTION

This module implements in PERL the functions that are currently part of the
Bitcoin Core RPC client calls (bitcoin-cli).The function names and parameters
are identical between the Bitcoin Core API and this module. This is done for
consistency so that a developer only has to reference one manual:
https://bitcoin.org/en/developer-reference#remote-procedure-calls-rpcs

=head1 CONSTRUCTOR

$btc = Bitcoin::RPC::Client->new( %options )

This method creates a new C<Bitcoin::RPC::Client> and returns it.

   Key                 Default
   -----------         -----------
   user                undef (Required)
   password            undef (Required)
   host                undef (Required)
   port                8332
   timeout             20
   ssl                 0
   verify_hostname     1
   syntax              0
   debug               0

verify_hostname - OpenSSL support has been removed from the Bitcoin Core 
project as of v0.12.0.

syntax - setting to 1 will turn on correct method name checking as well as 
usage errors. This works with all versions of Bitcoin Core, but the API class 
currently only contains what is valid for v0.12.0. You may want to keep this off
if you are using a version other than v0.12.0 and you are getting errors you 
think you should not be getting.

debug - Turns on raw HTTP request/response output from LWP::UserAgent.

=head1 AUTHOR

Wesley Hinds wesley.hinds@gmail.com

=head1 AVAILABILITY

The latest branch is avaiable from Github.

https://github.com/whindsx/Bitcoin-RPC-Client.git

=head1 LICENSE AND COPYRIGHT

Copyright 2016 Wesley Hinds.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut
